ruskeyverbs/views.py

```diff
@@ -26,19 +51,618 @@ def index(request):
+            return HttpResponseRedirect(reverse('arrange-words',
+                                                args=[pk]))
+    else:
+        raise Http404
+
+
+@login_required
+def ReproduceSentence(request, pk):
+    """
+    reproduce sentence quiz view
+    """
+    example_inst = get_object_or_404(Example, pk=pk)
+    form = ReproduceSentenceForm()
+    try:
+        request.session['quiz_state'] += 1
+        request.session.modified = True
+    except KeyError:
+        request.session['quiz_state'] = 0
+    if request.session['quiz_state'] == 4:
+        return render(request, 'ruskeyverbs/reproduce_sentence.html',
+                      context={'english_text': example_inst.translation_text,
+                               'form': form,
+                               'pk': pk,
+                               'file': example_inst.example_audio})
+    else:
+        raise Http404
+
+
+@login_required
+def ReproduceSentenceEval(request, pk):
+    example_inst = get_object_or_404(Example, pk=pk)
+    verb_pk = example_inst.verb.pk
+    verb = example_inst.verb
+    if request.method == 'POST':
+        form = ReproduceSentenceForm(request.POST)
+        if form.is_valid():
+            # fuzzy string comparison to examine the accuracy of user input
+            user_input = strip_punct_lower(form.cleaned_data['sentence_field'])
+            answer = strip_punct_lower(example_inst.russian_text)
+            score = fuzz.ratio(user_input, answer)
+            request.session['current_score'].append(score)
+            request.session['progress'] += round(1/16, 3)*100
+            if request.session['progress'] > 95:
+                request.session['progress'] = 100
+            """
+            calculate the average score from the session list (this is what
+            gets written to the PerformancePerExample model)
+            """
+            average_score = sum(
+                request.session['current_score'])/len(
+                    request.session['current_score'])
+            average_score = round(average_score, 2)
+            average_score /= 100
+            try:
+                request.session['quiz_summary'].append(
+                    (
+                        example_inst.russian_text,
+                        example_inst.translation_text,
+                        request.session['current_score'][0],
+                        request.session['current_score'][1],
+                        request.session['current_score'][2],
+                        score,
+                        int(average_score * 100)
+                    )
+
+                )
+            except KeyError:
+                request.session['quiz_summary'] = []
+            request.session.modified = True
+            """
+            test whether the user already has a PerformancePerExample for
+            this verb
+            """
+            try:
+                PerformanceObj = PerformancePerExample.objects.get(
```

Comment:
I feel like a lot of functionality like this block could and should be moved out of the views.py file and into the models themselves.

Reply:
Moved the PerformancePerExample update functionality into a class method.

Reply:
Good